ovn-tester/cms/openstack/openstack.py

```diff
@@ -0,0 +1,598 @@
+        self.external_port: Optional[LRPort] = None
+        self._projects: List[Project] = []
+
+        self._int_net_base = DualStackSubnet(cluster_cfg.node_net)
+        self._int_net_offset = 0
+
+        self._ext_net_pool_index = 0
+
+    @property
+    def projects(self) -> List[Project]:
+        """Return list of Projects that exist in Openstack deployment."""
+        return self._projects
+
+    def next_external_ip(self) -> DualStackIP:
+        """Return next available IP address from configured 'external_net'."""
+        self._ext_net_pool_index += 1
+        return self.cluster_cfg.external_net.forward(self._ext_net_pool_index)
+
+    def next_int_net(self):
+        """Return next subnet for project/tenant internal network.
+
+        This method takes subnet configured as 'node_net' in cluster config as
+        base and each call returns next subnet in line.
+        """
+        network = DualStackSubnet.next(
+            self._int_net_base, self._int_net_offset
+        )
+        self._int_net_offset += 1
+
+        return network
+
+    def new_project(self, gw_nodes: int = 0) -> Project:
+        """Create new Project/Tenant in the Openstack cloud.
+
+        Following things are provisioned for the Project:
+          * Project router
+          * Internal network
+
+        If 'gw_nodes' count is more than 0, external network is also
+        provisioned with default route through the gateway nodes. Gateway
+        nodes and their priority are selected at random from all
+        available 'worker_nodes'.
+
+        :param gw_nodes: Number of gateway chassis to use.
+        :return: New Project object that is automatically also added
+                 to 'self.projects' list.
+        """
+        project = Project()
+
+        project.router = self._create_project_router(
+            f"provider-router-{project.uuid}"
+        )
+
+        if gw_nodes:
+            self.add_external_network_to_project(
+                project,
+                self._get_gateway_chassis(gw_nodes),
+            )
+
+        self.add_internal_network_to_project(project, self.external_port)
+        self._projects.append(project)
+        return project
+
+    def add_external_network_to_project(
+        self, project: Project, gw_nodes: List[ChassisNode]
+    ) -> None:
+        """Provision external net to Project that adds external connectivity.
+
+        This method adds new network to project with prefix 'ext_net',
+        adds it to the Project's router and configures default route on the
+        router to go through this network.
+
+        :param project: Project to which external network will be added
+        :param gw_nodes: List of chassis that act as network gateways.
+        :return: None
+        """
+
+        ext_net = self._create_project_net(f"ext_net_{project.uuid}", 1500)
+        ext_net_port = self._add_metadata_port(ext_net, project.uuid)
+        provider_port = self._add_provider_network_port(ext_net)
```

Comment:
as far as i know there will only ever be one provider network port per provider network. Since `_add_provider_network_port` sets this statically to `physnet1` it should only ever be called once